src/server.rs

```diff
@@ -532,15 +530,114 @@ impl WalletService for Arc<crate::wallet::Wallet> {
         ))
     }
 
+    // Legacy Bitcoin Core-based implementation
+    // https://github.com/LayerTwo-Labs/mainchain/blob/05e71917042132202248c0c917f8ef120a2a5251/src/wallet/rpcwallet.cpp#L3863-L4008
     async fn create_bmm_critical_data_transaction(
         &self,
         request: tonic::Request<CreateBmmCriticalDataTransactionRequest>,
     ) -> std::result::Result<tonic::Response<CreateBmmCriticalDataTransactionResponse>, tonic::Status>
     {
-        Err(tonic::Status::new(
-            tonic::Code::Unimplemented,
-            "not implemented",
-        ))
+        let CreateBmmCriticalDataTransactionRequest {
+            sidechain_id,
+            value_sats,
+            height,
+            critical_hash,
+            prev_bytes,
+        } = request.into_inner();
+
+        let amount = value_sats
+            .ok_or_else(|| missing_field::<CreateBmmCriticalDataTransactionRequest>("value_sats"))
+            .map(bdk::bitcoin::Amount::from_sat)
+            .map_err(|_| {
+                invalid_field_value::<CreateBmmCriticalDataTransactionRequest>(
+                    "value_sats",
+                    &value_sats.unwrap_or_default().to_string(),
+                )
+            })?;
+
+        let locktime = height
+            .ok_or_else(|| missing_field::<CreateBmmCriticalDataTransactionRequest>("height"))
+            .map(bdk::bitcoin::absolute::LockTime::from_height)?
+            .map_err(|_| {
+                invalid_field_value::<CreateBmmCriticalDataTransactionRequest>(
+                    "height",
+                    &height.unwrap_or_default().to_string(),
+                )
+            })?;
+
+        let sidechain_number = sidechain_id
+            .ok_or_else(|| missing_field::<CreateBmmCriticalDataTransactionRequest>("sidechain_id"))
+            .map(SidechainNumber::try_from)?
+            .map_err(|_| {
+                invalid_field_value::<CreateBmmCriticalDataTransactionRequest>(
+                    "sidechain_id",
+                    &sidechain_id.unwrap_or_default().to_string(),
+                )
+            })?;
+
+        if !self
+            .is_sidechain_active(sidechain_number)
+            .await
+            .unwrap_or_default()
```

Comment:
If the server encounters an internal error, it is probably best not to ignore that

Reply:
agreed!